### DISCUSSION TITLE
🚀 Feature Request: Easy targetting of `.html` files in Cloudflare Pages `_headers` file

### DISCUSSION THREAD

**Discussion by Hexstream** [cloudflare/workers-sdk#13498] [Feature Requests]
### Describe the solution

Hello,

I am in the process of implementing [CSP](https://developer.mozilla.org/en-US/docs/Web/HTTP/CSP) across all my websites.

I therefore need to add a `Content-Security-Policy` header to all my HTML pages.

I expected this would be as simple as:

```
/*.html
  Content-Security-Policy: [...]
```

_Unfortunately, this does not work, and any workarounds seem very painful._

**I think this should simply target all `.html` files, including any `index.html`, `foo.html` and `404.html` files.**

_Could you please make this work as expected?_

**Comment by itsmatteomanf** [comment]
I think here the easier thing would be to apply it to all paths and eventually remove it or override it for paths in which it's not required (do note that it won't override it if you don't clear it before).

```
/*
  Content-Security-Policy: default-src 'self';

# This is an example, just to change something. It applies to any other header.
/assets/*
  !Content-Security-Policy
  Content-Security-Policy: default-src 'none';
```

The matching is done at the path level for the request, so the `404.html` file will never be actually hit directly and it would take whatever the path you have 404'd on had set to take. `index.html`, `foo.html`, etc. will always be redirected to the pretty URL (https://developers.cloudflare.com/pages/configuration/serving-pages/#route-matching) so you need to do as I'm showing above.

Adding the CSP header to other non-HTML assets it's actually not an issue.

**Comment by Hexstream** [comment]
> I think here the easier thing would be to apply it to all paths and eventually remove it or override it for paths in which it's not required

That's just _ludicrous._ I am simply not going to do that.

I can't believe I have to state this explicitly, but **_targetting `*.html` files is extremely useful in general_**.
I just happened to cite `Content-Security-Policy` as one example, since that's what I'm currently working on.

> The matching is done at the path level for the request [...]

That may well be how things _currently work_, but Cloudflare Pages' entire reason for existing is _making things easier for developers_.

> Adding the CSP header to other non-HTML assets it's actually not an issue.

_**It is absolutely an issue,**_ and I am categorically never going to do that, if only because **_`Content-Security-Policy` is a header that can get pretty big even in super simple scenarios,_** therefore sending it even when it is 100% irrelevant is, _especially at scale, **downright unconscionable.**_

The super simple CSP header I am working on is already 368 bytes long:

```
Content-Security-Policy: default-src 'none'; script-src 'self' https://global.hexstream.dev/scripts/arrows-madness.mjs https://global.hexstream.dev/scripts/dom.mjs; style-src 'self' https://global.hexstream.dev/css/; img-src 'self' https://global.hexstream.dev/; font-src https://global.hexstream.dev/fonts/; base-uri 'none'; form-action 'none'; frame-ancestors 'none'
```

**_There is simply NO WAY I am sending this header in billions of responses that inherently don't need it._**

**Comment by Hexstream** [comment]
Hilariously, I briefly tried locally adding a CSP to all files, just for kicks,
and **_immediately found out this can break chrome in unexpected ways:_**

![csp-applied](https://github.com/cloudflare/workers-sdk/assets/13000/6354c4aa-e460-49bc-b9ec-a7c34bdb779d)

![csp-not-applied](https://github.com/cloudflare/workers-sdk/assets/13000/2c12c2d7-46f3-429b-a63c-97a8d50fdf8e)

**Comment by itsmatteomanf** [comment]
But no other service allows for specific targeting of `*.html` files, unless the name appears in the path, which Pages doesn't support.

Also, 368 bytes (before compression, etc.) is nothing.

If you follow best practices all static assets are in a specific folder, as are images. It's literally 2 additional rules to remove that header (and you'll need the rule anyway if you want to have `Cache-Control` headers for the users that differ as they should with hashed and immutable files)...

The breakage in Chrome simply means they are not applying the centering styles that Chrome auto applies, but the image is displayed just fine.

**Comment by Hexstream** [comment]
> But no other service allows [...]

Sounds like a nice opportunity to outshine all those other services, then...

> Also, 368 bytes (before compression, etc.) is nothing.

It's not nothing on the scale of _billions (or even **trillions**, at Cloudflare scale) of responses_,
and with your kind of attitude I feel we wouldn't be sloppy _just this one time._

> If you follow best practices

Sorry, I don't believe I have anything to learn from you regarding best practices.

> The breakage in Chrome simply means they are not applying the centering styles that Chrome auto applies, but the image is displayed just fine.

Translation: "_You just immediately found **the tip of an iceberg,** which looks pretty small to me on the surface._" 🤦

-----

Anyway, _I am done arguing with you._ If your next reply is _again_ completely unhelpful, I am blocking you.

**Comment by itsmatteomanf** [comment]
First, I'm just a user trying to help you out finding a practical and quick solution, if you just bothered to look.

Second, putting static assets in a specific path is basic web design 101, it's helpful for *tons* of reasons. Including this.

Sure CSP on a specific static asset isn't the best, agreed, but it won't break everything 99.99% of the times. There are *really* easy solutions, it's literally one line in the same config file.
Would a match on the `Content-Type` header be better? Sure! Is it there? No. Will it be there soon? Probably not, and this is not even the right place to ask, either go to the Community or Discord.

**Comment by Hexstream** [comment]
> First, I'm just a user trying to help you out finding a practical and quick solution

I seriously doubt that, in any case _you are failing 100%._

The only _sane_ solution here is the great and widely useful feature I am proposing.

> Second, putting static assets in a specific path is basic web design 101

Again, I have nothing to learn from you regarding best practices.

**_Also note that 100% of my Cloudflare Pages files are static._**

> Sure CSP on a specific static asset isn't the best

It is _completely broken_ both [in theory](#issuecomment-2041574529) and [in practice](#issuecomment-2041582634).

> Would a match on the `Content-Type` header be better? Sure! Is it there? No. Will it be there soon? Probably not, and this is not even the right place to ask, either go to the Community or Discord.

What the fuck? **My suggestion is _literally 100% unrelated_ to the `Content-Type` header.**
The best place for my suggestion is obviously _here_, especially now that I'll never have to interact with you _ever again._

Cloudflare Pages already inherently needs to find and process all built `.html` files during deployment, including any `index.html`, `foo.html` and `404.html` files. Matching would be done on that basis only, obviously.

-----

You are being **completely unhelpful** for the _third time in a row._ [I warned you.](#issuecomment-2041629617) Cya.

**Comment by jasnell** [comment]
@Hexstream ... I'm not going to weigh in on the details of this specific request... I'll leave that the workers-sdk team. What I do want to ask is that you dial back quite a bit on the unnecessary anger you're showing here in your responses. You may disagree with @itsmatteomanf's suggestion but it's possible to do so respectfully and professionally and you seem to have failed to do either.

The problematic comments are in:

  * https://github.com/cloudflare/workers-sdk/issues/5546#issuecomment-2041629617
  * https://github.com/cloudflare/workers-sdk/issues/5546#issuecomment-2043700840

Refs: https://github.com/cloudflare/workers-sdk/blob/main/CODE_OF_CONDUCT.md

**Comment by petebacondarwin** [comment]
@Hexstream - I have hidden the comment above where you swear and are generally offensive to the user who had been trying to assist you. Please do consider your language when interacting on our repository and follow our Code of Conduct as @jasnell points out. We do not tolerate that kind of interaction here.
